src/librustc_mir/interpret/validity.rs

```diff
@@ -469,6 +480,13 @@ impl<'rt, 'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>>
                         ).is_ok() ||
                         self.ecx.memory.get_fn(ptr).is_ok();
                     if !non_null {
+                        if self.ref_tracking_for_consts.is_some() &&
+                            self.ecx.memory.get(ptr.alloc_id).is_err() &&
+                            self.ecx.memory.get_fn(ptr).is_err() {
+                            return validation_failure!(
+                                "encountered dangling pointer", self.path
+                            );
+                        }
                         // could be NULL
```

Comment:
This comment should be above the `if`, and it should also explain that the conditional is just to tweak the error message.

I am not sure if having a different error for OOB dangling pointers is worth it, but well. (This should not be hit for "just" dangling pointers, they must be OOB as well.)